File: Detect Cycle using DSU - GFG/detect-cycle-using-dsu.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
    public:
    
    int find(int parent[], int x){
        if(x==parent[x]) return x;
        return find(parent, parent[x]);
    }
    bool unite(int a, int b, int parent[], int size[]){
        a = find(parent,a);
        b = find(parent,b);
        if(a==b){
            return false;
        }
        if(size[a] > size[b]){
            parent[b] = a;
            size[a] += size[b];
        }else{
            parent[a] = b;
            size[b] += size[a];
        }
        return true;
    }
	int detectCycle(int V, vector<int>adj[])
	{
	    // make set
	    int parent[V]; for(int i = 0; i<V; i++) parent[i] = i;
	    int size[V]; for(int i = 0; i<V; i++) size[i] = 1;
	    // for each edge in graph
	    for(int i = 0; i<V; i++){
	        for(int j : adj[i]){
	            if(j>i){
	                if(!unite(i,j,parent,size)) return 1;
	            }
	        }
	    }
	 return 0;
	}
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int V, E;
		cin >> V >> E;
		vector<int>adj[V];
		for(int i = 0; i < E; i++){
			int u, v;
			cin >> u >> v;
			adj[u].push_back(v);
			adj[v].push_back(u);
		}
		Solution obj;
		int ans = obj.detectCycle(V, adj);
		cout << ans <<"\n";	}
	return 0;
}
// } Driver Code Ends

File: Detect Cycle using DSU - GFG/detect-cycle-using-dsu.js
//{ Driver Code Starts
//Initial Template for javascript


'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let [n,m] = readLine().trim().split(" ").map((x) => parseInt(x));
        let adj = [];
        for(let i = 0;i<n;i++){
            adj[i] = new Array();
        }
        
        for(let j = 0;j<m;j++){
            let [u,v] =  readLine().trim().split(" ").map((x) => parseInt(x));
            adj[u].push(v);
            adj[v].push(u);
        }
        let obj = new Solution();
        let res = obj.detectCycle(n,adj);
        console.log(res);
    }
}

// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number} n
 * @param {number[][]} adj
 * @return {number} 
*/

class Solution {
    detectCycle(n,adj){
       // makeset 
       const parent = new Array(n);
       for(let i = 0; i<parent.length; i++) parent[i] = i;
       const size = new Array(n).fill(1);
       // find representative
       const find = (x) =>{
           if(x==parent[x]) return x;
           return find(parent[x]);
       }
       const unite = function (a,b){
           a = find(a);
           b = find(b);
           if(a==b){
               return true;
               // containe cycle
           }
           if(size[a] > size[b]){
               parent[b] = a;
               size[a] += size[b];
           }else{
               parent[a] = b;
               size[b] += size[a];
           }
           return false;
       }
        for(let i = 0; i<n; i++){
            for(let j = 0; j<adj[i].length; j++){
                const adj_vertex = adj[i][j];
                if(adj_vertex > i){
                    if(unite(i,adj_vertex)) return 1;
                }
            }
        }
    return 0;
    }
}
